### ISSUE TITLE
Global Attribute "Type" Setting

### ISSUE THREAD

**Issue by WPprodigy** [woocommerce/woocommerce#18045]
I find this setting causes a lot of trouble in support when people manage to manually change it with really knowing what it does:

![type](https://user-images.githubusercontent.com/8536129/33681445-2e06a438-da8a-11e7-91bb-d681ccbfc213.png)

First part of confusion is them thinking it applies to the front-end, allowing users to enter a text field. I guess they didn't fully read the description, but I've definitely seen that come up a few times.

Second part of confusion used to be that the field just wouldn't save if it didn't contain a valid attribute, though after testing it seems terms are just created now if they don't already exist.

So the difference between select / text in terms of UX seems to be that with text you can create new global attributes quickly. With select, there is an "Add New" button that I just found for the first time: http://cld.wthms.co/UqUKRV. So perhaps if the flow for adding a new term with the select option was a bit smoother, there would be no need at all for text?

tl;dr - I'm opening this up to see if there would be a way to remove this setting entirely :)

**Comment by claudiosanches** [comment]
> Choose the Type. This is used for products. If you want to have the same attribute values for your whole store, choose “Select.” On the individual product, you can then select attributes. If you choose “Text”, you can type new values to the attributes.

Source: https://docs.woocommerce.com/document/managing-product-taxonomies/#set-global-attributes

@mikejolley do you remember something else about it?

**Comment by mikejolley** [comment]
Just allows free input which is faster in some cases I guess.

**Comment by mikejolley** [comment]
@kellychoffman @jameskoster Thoughts on removing?

**Comment by claudiosanches** [comment]
So better just keep "Select".

**Comment by kellychoffman** [comment]
+1 to remove. I had to read this 4 times to understand what the use was for, let's not make our customers do the same.

**Comment by mikejolley** [comment]
Looks like some plugins use `product_attributes_type_selector` so we can leave the option but remove 'text'.

**Comment by mikejolley** [comment]
Raised here https://github.com/woocommerce/woocommerce/pull/18071

**Comment by c-o-m-m-a-n-d-e-r** [comment]
Really? This destroyed a lot of work at my site.

can i bring this back with some hooks / filters ??

**Comment by mikejolley** [comment]
@c-o-m-m-a-n-d-e-r it literally only changes the input method in admin, nothing on frontend. You can still add new attributes inline using the 'add new' button in admin when editing a product.

**Comment by c-o-m-m-a-n-d-e-r** [comment]
yes i thought so but its not working.
i add Text back wirh filter, my options are in the DB ... but when i add to a product or want to change the input field didn‘t come... its just Title wirh empty value without a text input

selects are working

**Comment by mikejolley** [comment]
> yes i thought so but its not working.

remove your customisations and explain what is not working.

**Comment by c-o-m-m-a-n-d-e-r** [comment]
we were using Text and Select Attributes ... after the last Update(s) the Text Option was gone ...
But in the Database the Attributes are correctly saved with Type "text" ...

I can't edit existing Text Types now! I can add new correct ... but editing existing isn't possible ...

**Comment by WPprodigy** [comment]
Note that if you find yourself creating new terms for every product, you probably shouldn't be using global attribute - and product level attributes would be more fitting: http://cld.wthms.co/n5O4PL

**Comment by itsgauravjain22** [comment]
Suppose I want isbn13, isbn10, bookauthor, bookpulisher as a text attribute in each product. Your suggested method is to create these attributes each time when you add a product. That's highly inconvenient.

Also what if while if someone mispelled entering isbn13 to isb13, the other rest API client expect isbn13 field to be present. That's highly data error prone.

Removing a text type functionality from global attributes is highly inconvenient.

**Comment by mikejolley** [comment]
> Suppose I want isbn13, isbn10, bookauthor, bookpulisher as a text attribute in each product. Your suggested method is to create these attributes each time when you add a product. That's highly inconvenient.

Create a draft product with these fields. 

When you want a new one, duplicate it, edit the values, publish.

**Comment by itsgauravjain22** [comment]
Thanks for the trick. But It's not a suitable way to enforce a field name.

**Comment by WPprodigy** [comment]
Hmm, so Author/Publisher makes sense as a global taxonomy since there can be multiple products that share the same values - the purpose of taxonomies. There is value in seeing all books belonging to a specific author.

But isbn is unique to a specific product - so having a term for that does seem like a mistake in architecture. I'd not recommend going with a bad architecture decision for the sake of easiness in UI.

> Also what if while if someone mispelled entering isbn13 to isb13, the other rest API client expect isbn13 field to be present. That's highly data error prone.

This could be programmatically protected against when a product is saved perhaps?

**Comment by itsgauravjain22** [comment]
@WPprodigy That's what I found lack in default woocommerce. The custom text field can't be added as product attribute. It's just a example of books as a product. If there's a way to create custom text fields as products attributes, I could add isbn13, isbn10, upc, ean, total number of pages in a book etc. And it's very much required for every other store type. Not all product attributes are taxonomies which the woocommerce architecture enforces. I hope woocommerce team provides the store owner the flexibility of text fields as product attributes in future in efficient architecture.

**Comment by WPprodigy** [comment]
Right but the thing is - global attributes is the wrong approach for data like that. It's data that only relates to a single object, there's no need for it to be global.

This would be similar to creating a new category for every post you create on a news site.

**Comment by itsgauravjain22** [comment]
The product attributes data has two parts. It's name and it's vales. The value part is unique for each product (single object you referred). But the name part can be global, which I believe text product attributes are useful for. Handling of such text type product attributes could be extended in other way that woocommerce currently handles multiselect global product attributes. It's just like we're registering a sku field but as a named global text type product attribute whose name is defined and values are just unrelated text.

**Comment by WPprodigy** [comment]
A global attribute is it's own taxonomy (think categories, tags, etc). And the individual names are terms. You can't do one without the other. Creating a new taxonomy just to have a standardized field title is the architectural problem I'm referring too.

Instead, this sort of value (isbn) should likely just be postmeta - connected to the one object. Could have a custom field anywhere in the admin you need where this data is input, and then you can output it if it exists in the templates: https://github.com/woocommerce/woocommerce/blob/a5d62b83a87cf38f7e5a132a50a9bd0641315414/templates/single-product/product-attributes.php#L26-L33

**Comment by itsgauravjain22** [comment]
Thanks @WPprodigy to point in the correct direction. `post_meta` is the term I should use in previous comments. Does it makes sense if woocommerce have graphical interface to register new `post_meta` when user creates text type global product attribute.

**Comment by WPprodigy** [comment]
I don't think it's a great place for WC core to handle - namely it adds more confusion to the attribute landscape.

To a degree, I'd imagine most shops can just rely on the correct title being entered for custom product attributes. After all, there is nothing stopping a typo from being entered into the value (ISBN number). And with CSV exports/imports this could be extra checked over.

If a store scales up to a very large catalog and has specific needs, then it's time to do custom fields and automations where it makes sense.

That's just my two cents though :)

**Comment by itsgauravjain22** [comment]
Yes you're right that typo won't affect low sku store. But for high sku store it's a pain. The main reason behind having fixed named text products attribute to have consistent data over wc Rest API. The rest API client need to know about product attributes names. And usually attributes names are not modified, but their values are.

**Comment by erkange** [comment]
so do we have real solution for this case.
I want to add brand, EAN, Model name to all products.
I also want to show it to customer in front end.
I also want the to see in comparison table.

**Comment by KoenraadMoreau** [comment]
We also need to add : 
1° Brand
2° Translator
3° Illustrator
4° ISBN
5° Number of pages
It should be useful.
There is a plugin for that, but its not released since the beginning : https://wordpress.org/plugins/woo-load-group-attributes-product/
